README.md
WP Mail From II
===============

Allows you to configure the default email address and name used for emails sent by WordPress.

This plugin allows you to set the email address and name used for emails sent by WordPress by setting the *From:* header.

It is an updated and fully re-worked version of the [WP Mail From](http://wordpress.org/extend/plugins/wp-mailfrom/) plugin by Tristan Aston and now works with the latest versions of WordPress.

* Adds a "Mail From" section in the "Settings" menu.
* The plugin uses the filter hooks `wp_mail_from` and `wp_mail_from_name`.
* The priority for the hooks is set to 1 to allow for other plugins that may hook these with the default priority of 10 to override this plugin.

Danish translation by Frederik Svarre ([www.fslab.dk](http://www.fslab.dk/))

Installation
------------

Either install via the WordPress admin plugin installer or...

1. Unzip `wp-mailfrom-ii.zip` in the `/wp-content/plugins/` directory, making sure the folder is called `wp-mailfrom-ii`.
1. Activate the plugin through the 'Plugins' menu in WordPress.
1. Visit the admin settings page `Settings > Mail From` and save your preferred name and email address.

Upgrading from the old WP Mail From plugin
------------------------------------------

This version is pretty much a complete re-write, fixes loads of bugs and works with the most recent versions of WordPress.

If upgrading from the [WP Mail From](http://wordpress.org/extend/plugins/wp-mailfrom/) plugin your current name an email settings should be copied across but please check.
To upgrade simply activate this plugin then deactivate the old WP Mail From plugin.

Option naming conventions have changed in this version of the plugin.
If you accessed either of the options directly for any reason you will need to re-factor you code as support for the old options will be removed in a future version.

The old options used to be retrieved as follows:

`get_option( 'site_mail_from_name' );
get_option( 'site_mail_from_email' );`

You should now use:

`get_option( 'wp_mailfrom_ii_name' );
get_option( 'wp_mailfrom_ii_email' );`

Frequently Asked Questions
--------------------------

__Why does the From address still show as the default or show up as 'sent on behalf of' the default address?__  
Possibly your mail server has added a *Sender:* header or is configured to always set the *envelope sender* to the user calling it.

__Why are emails not being sent?__  
Some hosts may refuse to relay mail from an unknown domain. See [http://trac.wordpress.org/ticket/5007](http://trac.wordpress.org/ticket/5007) for more details.

Upgrade Notice
--------------

### 2.0
Add default values as placeholders in admin fields and new method of getting sitename in WordPress 5.5+

### 1.2
Makes admin field labels clickable and removes deprecated `screen_icon()`.

### 1.1.1
Update translations and bumped WordPress requirements.

### 1.1
Added options to override default WordPress email addresses and added 'wp_mailfrom_ii_default_from' filter.

### 1.0.2
Only set email address and name if overwriting the default WordPress values.

### 1.0.1
Correctly style save settings button.

### 1.0
This version is pretty much a complete re-write, fixes loads of bugs and works with the most recent versions of WordPress.
If upgrading from the [WP Mail From](http://wordpress.org/extend/plugins/wp-mailfrom/) plugin your current name an email settings should be copied across but please check.

Changelog
---------

View a list of all plugin changes in [CHANGELOG.md](https://github.com/benhuson/wp-mailfrom/blob/master/CHANGELOG.md).


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [2.0] - 2020-12-14

### Added
- Add default values as placeholders in admin fields.

### Changed
- New method of getting sitename in WordPress 5.5+

## [1.2] - 2020-12-14

### Added
- Make admin field labels clickable. Props [Viktor Szépe](https://github.com/szepeviktor).

### Fixed
- Remove deprecated `screen_icon()`.

## [1.1.1] - 2019-11-07

### Added
- Added Hungarian translation. Props [Viktor Szépe](https://github.com/szepeviktor).

### Changed
- Bump WordPress requirements.

## [1.1] - 2015-10-29

### Added
- Added options to override the default WordPress email address and admin email address.
- Added 'wp_mailfrom_ii_default_from' filter so you can add compatibility if the pluggable wp_mail() function is altered to use a different default email address.
- Delete plugin options when uninstalled.

### Changed
- Reworked as a singleton class.

### Removed
- Remove filter support for original WP MailFrom plugin.

## [1.0.2] - 2013-11-20

### Fixed
- Only set email address and name if overwriting the default WordPress values.

## [1.0.1] - 2012-11-29

### Fixed
- Correctly style "save settings" button.

## [1.0] - 2012-08-22

### Changed
- Pretty much re-coded from scratch - now based around a core WP_MailFrom_II class.
- Uses the [WordPress Settings API](http://codex.wordpress.org/Settings_API).
- Stores name and email as `wp_mailfrom_ii_name` and `wp_mailfrom_ii_email` options. Upgrade support is provided for old options.

## [0.6] - 2012-07-27

### Added
- Fork of original [WP Mail From](https://wordpress.org/plugins/wp-mailfrom/) plugin

[Unreleased]: https://github.com/benhuson/wp-mailfrom/compare/2.0...HEAD
[2.0]: https://github.com/benhuson/wp-mailfrom/compare/1.2...2.0
[1.2]: https://github.com/benhuson/wp-mailfrom/compare/1.1.1...1.2
[1.1.1]: https://github.com/benhuson/wp-mailfrom/compare/1.1...1.1.1
[1.1]: https://github.com/benhuson/wp-mailfrom/compare/1.0.2...1.1
[1.0.2]: https://github.com/benhuson/wp-mailfrom/compare/1.0.1...1.0.2
[1.0.1]: https://github.com/benhuson/wp-mailfrom/compare/1.0...1.0.1
[1.0]: https://github.com/benhuson/wp-mailfrom/compare/0.6...1.0
[0.6]: https://github.com/benhuson/wp-mailfrom/releases/tag/0.6


wp-mailfrom.php
<?php 

/*
Plugin Name: WP Mail From II
Version: 1.0
Plugin URI: http://wordpress.org/extend/plugins/wp-mailfrom-ii/
Description: Allows you to configure the default email address and name used for emails sent by WordPress.
Author: Ben Huson

An updated and fully re-worked version of the WP Mail From plugin by Tristan Aston.
http://wordpress.org/extend/plugins/wp-mailfrom/

Copyright (c) 2012, 
Released under the GPL license
http://www.gnu.org/licenses/gpl.txt
*/

class WP_MailFrom_II {

	/**
	 * Constructor
	 */
	function WP_MailFrom_II() {
		add_action( 'admin_init', array( $this, 'settings' ) );
		add_action( 'admin_menu', array( $this, 'admin_menu' ) );
		add_action( 'plugins_loaded', array( $this, 'load_textdomain' ) );
		
		// Name and email filter
		add_filter( 'wp_mail_from_name', array( $this, 'wp_mail_from_name' ), 1 );
		add_filter( 'wp_mail_from', array( $this, 'wp_mail_from' ), 1 );
		
		// Legacy support for old options - just in case someone used this directly!
		// (not really needed unless we can takeover the old plugin)
		//add_filter( 'pre_option_site_mail_from_name', 'get_option_site_mail_from_name', 1 );
		//add_filter( 'pre_option_site_mail_from_email', 'get_option_site_mail_from_email', 1 );
	}
	
	/**
	 * Load Text Domain Language Support
	 */
	function load_textdomain() {
		load_plugin_textdomain( 'wp-mailfrom-ii', false, dirname( plugin_basename( __FILE__ ) ) . '/lang/' );
	}
	
	/**
	 * Filter: wp_mail_from_name
	 *
	 * @param string $name Default name.
	 * @return string $name WP Mail From name.
	 */
	function wp_mail_from_name( $name ) {
		$wp_mailfrom_ii_name = get_option( 'wp_mailfrom_ii_name', '' );
		if ( ! empty( $wp_mailfrom_ii_name ) ) {
			return $wp_mailfrom_ii_name;
		}
		return $name;
	}
	
	/**
	 * Filter: wp_mail_from
	 *
	 * @param string $name Default email.
	 * @return string $name WP Mail From email.
	 */
	function wp_mail_from( $email ) {
		$wp_mailfrom_ii_email = get_option( 'wp_mailfrom_ii_email', '' );
		if ( ! empty( $wp_mailfrom_ii_email ) ) {
			return $wp_mailfrom_ii_email;
		}
		return $email;
	}
	
	/**
	 * Admin Menu
	 */
	function admin_menu() {
		add_options_page( __( 'WP Mail From Plugin', 'wp-mailfrom-ii' ), __( 'Mail From', 'wp-mailfrom-ii' ), 'manage_options', 'wp_mailfrom', array( $this, 'settings_page' ) );
	}
	
	/**
	 * Settings Page
	 */
	function settings_page() {
		?>
		<div class="wrap">
			<div id="icon-options-general" class="icon32"><br></div>
			<h2><?php _e( 'Mail From Settings', 'wp-mailfrom-ii' ); ?></h2>
			<form action="options.php" method="post">
				<?php
				settings_fields( 'wp_mailfrom_ii' );
				do_settings_sections( 'wp_mailfrom_ii' );
				?>
				<p class="submit"><input name="submit" type="submit" value="<?php esc_attr_e( 'Save Changes', 'wp-mailfrom-ii' ); ?>" /></p>
			</form>
		</div>
		<?php
	}
	
	/**
	 * Settings API
	 */
	function settings() {
		add_settings_section(
			'wp_mailfrom_ii',
			'',
			array( $this, 'settings_section' ),
			'wp_mailfrom_ii'
		);
		add_settings_field(
			'wp_mailfrom_ii_name',
			__( 'From Name', 'wp-mailfrom-ii' ),
			array( $this, 'wp_mailfrom_ii_name_field' ),
			'wp_mailfrom_ii',
			'wp_mailfrom_ii'
		);
		add_settings_field(
			'wp_mailfrom_ii_email',
			__( 'From Email Address', 'wp-mailfrom-ii' ),
			array( $this, 'wp_mailfrom_ii_email_field' ),
			'wp_mailfrom_ii',
			'wp_mailfrom_ii'
		);
 		register_setting( 'wp_mailfrom_ii', 'wp_mailfrom_ii_name', array( $this, 'sanitize_wp_mailfrom_ii_name' ) );
 		register_setting( 'wp_mailfrom_ii', 'wp_mailfrom_ii_email', 'is_email' );
	}

	/**
	 * Sanitize Mail From Name
	 * Strips out all HTML, scripts...
	 *
	 * @param string $val Name.
	 * @return string Sanitized name.
	 */
	function sanitize_wp_mailfrom_ii_name( $val ) {
		return wp_kses( $val, array() );
	}

	/**
	 * Mail From Settings Section
	 */
	function settings_section() {
		echo '<p>' . __( 'If set, these 2 options will override the name and email address in the <strong>From:</strong> header on all sent emails.', 'wp-mailfrom-ii' ) . '</p>';
	}

	/**
	 * Mail From Name Field
	 */
	function wp_mailfrom_ii_name_field() {
		echo '<input name="wp_mailfrom_ii_name" type="text" id="wp_mailfrom_ii_name" value="' . get_option( 'wp_mailfrom_ii_name', '' ) . '" class="regular-text" />';
	}

	/**
	 * Mail From Email Field
	 */
	function wp_mailfrom_ii_email_field() {
		echo '<input name="wp_mailfrom_ii_email" type="text" id="wp_mailfrom_ii_email" value="' . get_option( 'wp_mailfrom_ii_email', '' ) . '" class="regular-text" />';
	}
	
	/**
	 * Legacy support for get_option( 'site_mail_from_name' )
	 */
	function get_option_site_mail_from_name( $option, $default = false ) {
		return get_option( 'wp_mailfrom_ii_name', $default );
	}
	
	/**
	 * Legacy support for get_option( 'site_mail_from_email' )
	 */
	function get_option_site_mail_from_email( $option, $default = false ) {
		return get_option( 'wp_mailfrom_ii_email', $default );
	}
	
	/**
	 * Register Activation
	 * Perform upgrades etc.
	 */
	function register_activation() {
		
		// Temporarily remove our filter which provide support for legacy options
		// (is only really needed if we can takeover the old plugin, but leave in for now)
		remove_filter( 'pre_option_site_mail_from_name', 'get_option_site_mail_from_name', 1 );
		remove_filter( 'pre_option_site_mail_from_email', 'get_option_site_mail_from_email', 1 );
		
		// Get old option value and try to assign them to new options
		$name = get_option( 'site_mail_from_name', '' );
		$email = get_option( 'site_mail_from_email', '' );
		$new_name = get_option( 'wp_mailfrom_ii_name', '' );
		$new_email = get_option( 'wp_mailfrom_ii_email', '' );
		if ( ! empty( $name ) && empty( $new_name ) )
			$name_updated = add_option( 'wp_mailfrom_ii_name', $name );
		if ( ! empty( $email ) && empty( $new_email ) )
			$email_updated = add_option( 'wp_mailfrom_ii_email', $email );
		
		// If new options created delete old options
		// (don't do this at the moment, only if we can takeover the old plugin)
		/*
		if ( $name_updated )
			delete_option( 'site_mail_from_name' );
		if ( $email_updated )
			delete_option( 'site_mail_from_email' );
		*/
	}
	
}

global $WP_MailFrom_II;
$WP_MailFrom_II = new WP_MailFrom_II();
register_activation_hook( __FILE__, array( $WP_MailFrom_II, 'register_activation' ) );


wp-mailfrom-ii.php
<?php 

/**
 * Plugin Name:        WP Mail From II
 * Plugin URI:         https://wordpress.org/plugins/wp-mailfrom-ii/
 * Description:        Allows you to configure the default email address and name used for emails sent by WordPress.
 * Version:            2.0
 * Author:             Ben Huson
 * Author URI:         https://github.com/benhuson/
 * Minimum WordPress Version Required: 4.3
 * Tested up to:       5.6
 * Text Domain:        wp-mailfrom-ii
 * Domain Path:        /languages
 * License:            GPL-2.0+
 * License URI:        http://www.gnu.org/licenses/gpl-2.0.txt
 * GitHub Plugin URI:  https://github.com/benhuson/wp-mailfrom
 *
 * An updated and fully re-worked version of the WP Mail From plugin by Tristan Aston.
 * http://wordpress.org/extend/plugins/wp-mailfrom/
 */

if ( ! defined( 'ABSPATH' ) ) exit;  // Exit if accessed directly

// Require public files.
require_once( plugin_dir_path( __FILE__ ) . 'public/class-wp-mailfrom-ii.php' );

// Register hooks that are fired when the plugin is activated.
register_activation_hook( __FILE__, array( 'WP_MailFrom_II', 'activate' ) );

// Init.
add_action( 'plugins_loaded', array( 'WP_MailFrom_II', 'get_instance' ) );

// Only load admin functionality in admin.
if ( is_admin() && ( ! defined( 'DOING_AJAX' ) || ! DOING_AJAX ) ) {

	require_once( plugin_dir_path( __FILE__ ) . 'admin/class-wp-mailfrom-ii-admin.php' );
	add_action( 'plugins_loaded', array( 'WP_MailFrom_II_Admin', 'get_instance' ) );

}


admin/class-wp-mailfrom-ii-admin.php
<?php

/**
 * @package     WP Mail From II
 * @subpackage  Admin Class
 *
 * This class is used to work with the
 * administrative side of the WordPress site.
 */

if ( ! defined( 'ABSPATH' ) ) exit;  // Exit if accessed directly

class WP_MailFrom_II_Admin {

	/**
	 * Instance of this class.
	 *
	 * @since  1.1
	 *
	 * @var    object
	 */
	protected static $instance = null;

	/**
	 * Slug of the plugin screen.
	 *
	 * @since  1.1
	 *
	 * @var    string
	 */
	protected $plugin_screen_hook_suffix = null;

	/**
	 * Initialize the plugin by loading admin scripts & styles and adding a
	 * settings page and menu.
	 *
	 * @since  1.1
	 */
	private function __construct() {

		// Get plugin slug
		$plugin = WP_MailFrom_II::get_instance();
		$this->plugin_slug = $plugin->get_plugin_slug();

		// Setup default options
		add_action( 'admin_init', array( $this, 'setup_default_options' ) );

		// Add the options page and menu item.
		add_action( 'admin_menu', array( $this, 'add_plugin_admin_menu' ) );

		// Register settings fields
		add_action( 'admin_init', array( $this, 'settings' ) );

		// Add an action link pointing to the settings page.
		$this->plugin_basename = plugin_basename( plugin_dir_path( __DIR__ ) . $this->plugin_slug . '.php' );
		add_filter( 'plugin_action_links_' . $this->plugin_basename, array( $this, 'add_action_links' ) );
		add_filter( 'plugin_row_meta', array( $this, 'plugin_row_meta' ), 10, 4 );

	}

	/**
	 * Setup default options.
	 *
	 * @since  1.1
	 */
	public function setup_default_options() {

		add_option( 'wp_mailfrom_ii_override_default', 1 );
		add_option( 'wp_mailfrom_ii_override_admin', 1 );

	}

	/**
	 * Return an instance of this class.
	 *
	 * @since   1.1
	 *
	 * @return  object  A single instance of this class.
	 */
	public static function get_instance() {

		// If the single instance hasn't been set, set it now.
		if ( null == self::$instance ) {
			self::$instance = new self;
		}

		return self::$instance;

	}

	/**
	 * Register administration menus.
	 *
	 * @since  1.1
	 */
	public function add_plugin_admin_menu() {

		// Add a settings page to the Settings menu
		$this->plugin_screen_hook_suffix = add_options_page(
			__( 'WP Mail From Plugin', 'wp-mailfrom-ii' ),
			__( 'Mail From', 'wp-mailfrom-ii' ),
			'manage_options',
			$this->plugin_slug,
			array( $this, 'display_plugin_admin_page' )
		);

	}

	/**
	 * Render the settings page.
	 *
	 * @since  1.1
	 */
	public function display_plugin_admin_page() {

		include_once( 'views/admin.php' );

	}

	/**
	 * Settings API.
	 *
	 * @since  1.1
	 */
	public function settings() {

		add_settings_section(
			'wp_mailfrom_ii',
			'',
			array( $this, 'settings_section' ),
			'wp_mailfrom_ii'
		);
		add_settings_field(
			'wp_mailfrom_ii_name',
			__( 'From Name', 'wp-mailfrom-ii' ),
			array( $this, 'wp_mailfrom_ii_name_field' ),
			'wp_mailfrom_ii',
			'wp_mailfrom_ii',
			array( 'label_for' => 'wp_mailfrom_ii_name' )
		);
		add_settings_field(
			'wp_mailfrom_ii_email',
			__( 'From Email Address', 'wp-mailfrom-ii' ),
			array( $this, 'wp_mailfrom_ii_email_field' ),
			'wp_mailfrom_ii',
			'wp_mailfrom_ii',
			array( 'label_for' => 'wp_mailfrom_ii_email' )
		);
		add_settings_field(
			'wp_mailfrom_ii_override',
			__( 'Override Emails From', 'wp-mailfrom-ii' ),
			array( $this, 'wp_mailfrom_ii_override_fields' ),
			'wp_mailfrom_ii',
			'wp_mailfrom_ii'
		);

		register_setting( 'wp_mailfrom_ii', 'wp_mailfrom_ii_name', array( $this, 'sanitize_wp_mailfrom_ii_name' ) );
		register_setting( 'wp_mailfrom_ii', 'wp_mailfrom_ii_email', 'is_email' );
		register_setting( 'wp_mailfrom_ii', 'wp_mailfrom_ii_override_default', 'absint' );
		register_setting( 'wp_mailfrom_ii', 'wp_mailfrom_ii_override_admin', 'absint' );

	}

	/**
	 * Sanitize Mail From Name.
	 *
	 * Strips out all HTML, scripts, etc...
	 *
	 * @since   1.1
	 *
	 * @param   string  $val  Name.
	 * @return  string        Sanitized name.
	 */
	public function sanitize_wp_mailfrom_ii_name( $val ) {

		return wp_kses( $val, array() );

	}

	/**
	 * Mail From Settings Section.
	 *
	 * @since  1.1
	 */
	public function settings_section() {

		echo '<p>' . __( 'If set, these two options will override the default name and email address in the &quot;From&quot; header on emails sent by WordPress.', 'wp-mailfrom-ii' ) . '</p>';

	}

	/**
	 * Mail From Name Field.
	 *
	 * @since  1.1
	 */
	public function wp_mailfrom_ii_name_field() {

		$wp_mailfrom_ii = \WP_MailFrom_II::get_instance();

		echo '<input name="wp_mailfrom_ii_name" type="text" id="wp_mailfrom_ii_name" value="' . esc_attr( get_option( 'wp_mailfrom_ii_name', '' ) ) . '" class="regular-text" placeholder="' . esc_attr( $wp_mailfrom_ii->get_default_from_name() ) . '" />';

	}

	/**
	 * Mail From Email Field.
	 *
	 * @since  1.1
	 */
	public function wp_mailfrom_ii_email_field() {

		$wp_mailfrom_ii = \WP_MailFrom_II::get_instance();

		echo '<input name="wp_mailfrom_ii_email" type="text" id="wp_mailfrom_ii_email" value="' . esc_attr( get_option( 'wp_mailfrom_ii_email', '' ) ) . '" class="regular-text" placeholder="' . esc_attr( $wp_mailfrom_ii->get_default_from() ) . '" />';

	}

	/**
	 * Mail From Override Fields.
	 *
	 * @since  1.1
	 */
	public function wp_mailfrom_ii_override_fields() {

		$wp_mailfrom = WP_MailFrom_II::get_instance();
		$email = $wp_mailfrom->get_default_from();

		echo '<label><input name="wp_mailfrom_ii_override_default" type="checkbox" id="wp_mailfrom_ii_override_default" value="1"' . checked( 1, get_option( 'wp_mailfrom_ii_override_default', 0 ), false ) . ' /> ' . esc_html__( 'Default WordPress Email', 'wp-mailfrom-ii' ) . ' <span class="description">(' . esc_html( $email ) . ')</span></label><br />';
		echo '<label><input name="wp_mailfrom_ii_override_admin" type="checkbox" id="wp_mailfrom_ii_override_admin" value="1"' . checked( 1, get_option( 'wp_mailfrom_ii_override_admin', 0 ), false ) . ' /> ' . esc_html__( 'Admin Email', 'wp-mailfrom-ii' ) . ' <span class="description">(' . esc_html( get_option( 'admin_email' ) ) . ')</span></label>';

	}

	/**
	 * Add settings action link to the plugins page.
	 *
	 * @since  1.1
	 */
	public function add_action_links( $links ) {

		return array_merge(
			array(
				'settings' => '<a href="' . admin_url( 'options-general.php?page=' . $this->plugin_slug ) . '">' . esc_html__( 'Settings', 'wp-mailfrom-ii' ) . '</a>'
			),
			$links
		);

	}

	/**
	 * Plugin Row Meta
	 *
	 * Adds documentation, support and issue links below the plugin description on the plugins page.
	 *
	 * @since   1.1
	 *
	 * @param   array   $plugin_meta  Plugin meta display array.
	 * @param   string  $plugin_file  Plugin reference.
	 * @param   array   $plugin_data  Plugin data.
	 * @param   string  $status       Plugin status.
	 * @return  array                 Plugin meta array.
	 */
	public function plugin_row_meta( $plugin_meta, $plugin_file, $plugin_data, $status ) {

		if ( $this->plugin_basename == $plugin_file ) {
			$plugin_meta[] = sprintf( '<a href="%s">%s</a>', 'https://github.com/benhuson/wp-mailfrom', esc_html__( 'GitHub', 'wp-mailfrom-ii' ) );
			$plugin_meta[] = sprintf( '<a href="%s">%s</a>', esc_url( __( 'http://wordpress.org/support/plugin/wp-mailfrom-ii', 'wp-mailfrom-ii' ) ), esc_html__( 'Support', 'wp-mailfrom-ii' ) );
		}

		return $plugin_meta;

	}

}


public/class-wp-mailfrom-ii.php
<?php

/**
 * @package     WP Mail From II
 * @subpackage  Public Class
 *
 * This class should be used to work with the
 * public-facing side of the WordPress site.
 */

if ( ! defined( 'ABSPATH' ) ) exit;  // Exit if accessed directly

class WP_MailFrom_II {

	/**
	 * Plugin version, used for cache-busting of style and script file references and db upgrades.
	 *
	 * @since  1.1
	 *
	 * @var    string
	 */
	const VERSION = '1.1';

	/**
	 * Unique identifier for your plugin.
	 *
	 * The variable name is used as the text domain when internationalizing strings
	 * of text. Its value should match the Text Domain file header in the main
	 * plugin file.
	 *
	 * @since  1.1
	 *
	 * @var    string
	 */
	protected $plugin_slug = 'wp-mailfrom-ii';

	/**
	 * Instance of this class.
	 *
	 * @since  1.1
	 *
	 * @var    object
	 */
	protected static $instance = null;

	/**
	 * Constructor
	 *
	 * @since  1.1
	 */
	private function __construct() {

		add_action( 'init', array( $this, 'load_plugin_textdomain' ) );

		// Name and email filter
		add_filter( 'wp_mail_from_name', array( $this, 'wp_mail_from_name' ), 100 );
		add_filter( 'wp_mail_from', array( $this, 'wp_mail_from' ), 100 );

	}

	/**
	 * Return an instance of this class.
	 *
	 * @since   1.1
	 *
	 * @return  object  A single instance of this class.
	 */
	public static function get_instance() {

		// If the single instance hasn't been set, set it now.
		if ( null == self::$instance ) {
			self::$instance = new self;
		}

		return self::$instance;

	}

	/**
	 * Return the plugin slug.
	 *
	 * @since   1.1
	 *
	 * @return  Plugin slug variable.
	 */
	public function get_plugin_slug() {

		return $this->plugin_slug;

	}

	/**
	 * Load Text Domain Language Support
	 *
	 * @since  1.1
	 */
	public function load_plugin_textdomain() {

		$plugin_basename = plugin_basename( plugin_dir_path( __DIR__ ) . $this->plugin_slug . '.php' );
		load_plugin_textdomain( 'wp-mailfrom-ii', false, dirname( $plugin_basename ) . '/languages/' );

	}

	/**
	 * Filter: wp_mail_from_name
	 *
	 * @since   1.1
	 *
	 * @param   string  $name  Default name.
	 * @return  string         WP Mail From name.
	 */
	public function wp_mail_from_name( $name ) {

		$wp_mailfrom_ii_name = get_option( 'wp_mailfrom_ii_name', '' );

		if ( ! empty( $wp_mailfrom_ii_name ) && ! $this->is_default_from_name( $wp_mailfrom_ii_name ) ) {
			return $wp_mailfrom_ii_name;
		}

		return $name;

	}

	/**
	 * Filter: wp_mail_from
	 *
	 * @since   1.1
	 *
	 * @param   string $name  Default email.
	 * @return  string        WP Mail From email.
	 */
	public function wp_mail_from( $email ) {

		$wp_mailfrom_ii_email = get_option( 'wp_mailfrom_ii_email', '' );

		if ( ! empty( $wp_mailfrom_ii_email ) && is_email( $wp_mailfrom_ii_email ) ) {

			$override_default = get_option( 'wp_mailfrom_ii_override_default', 0 );
			$override_admin = get_option( 'wp_mailfrom_ii_override_admin', 0 );

			if ( $override_default == 1 && $this->is_default_from( $email ) ) {
				return $wp_mailfrom_ii_email;
			}

			if ( $override_admin == 1 && $this->is_admin_from( $email ) ) {
				return $wp_mailfrom_ii_email;
			}

		}

		return $email;

	}

	/**
	 * Is Default From Name
	 *
	 * @since   1.1
	 *
	 * @param   string   $name  Name to check.
	 * @return  boolean
	 */
	public function is_default_from_name( $name ) {

		return $this->get_default_from_name() == $name;

	}

	/**
	 * Get Default From Name
	 *
	 * Get the default name assigned by WordPress.
	 * This is defined in wp_mail() in wp-includes/pluggable.php
	 *
	 * @since   2.0
	 *
	 * @return  string  Default name.
	 */
	public function get_default_from_name() {

		return 'WordPress';

	}

	/**
	 * Is Default From Email
	 *
	 * @since   1.1
	 *
	 * @param   string   $email  Email to check.
	 * @return  boolean
	 */
	public function is_default_from( $email ) {

		return $this->get_default_from() == $email;

	}

	/**
	 * Get Default From Email
	 *
	 * Checks to see if the email is the default address assigned by WordPress.
	 * This is defined in wp_mail() in wp-includes/pluggable.php
	 *
	 * The 'wp_mailfrom_ii_default_from' filter is provided so you can add compatibility when
	 * the pluggable wp_mail() function is altered to use a different default email address.
	 *
	 * Also note, some hosts may refuse to relay mail from an unknown domain. See
	 * http://trac.wordpress.org/ticket/5007
	 *
	 * @since   1.1
	 *
	 * @return  string  Default from email.
	 */
	public function get_default_from() {

		global $wp_version;

		/**
		 * New method of getting sitename in WordPress 5.5+
		 *
		 * @see  https://core.trac.wordpress.org/ticket/25239#comment:114
		 * @see  https://core.trac.wordpress.org/changeset/48601
		 */
		if ( version_compare( $wp_version, '5.5' ) >= 0 ) {
			$sitename = wp_parse_url( network_home_url(), PHP_URL_HOST );
		} else {
			$sitename = strtolower( $_SERVER['SERVER_NAME'] );
		}

		if ( 'www.' === substr( $sitename, 0, 4 ) ) {
			$sitename = substr( $sitename, 4 );
		}

		return apply_filters( 'wp_mailfrom_ii_default_from', 'wordpress@' . $sitename );

	}

	/**
	 * Is Admin From Email
	 *
	 * Checks to see if the email is the admin email address set in the WordPress options.
	 *
	 * Also note, some hosts may refuse to relay mail from an unknown domain. See
	 * http://trac.wordpress.org/ticket/5007
	 *
	 * @since   1.1
	 *
	 * @param   string   $email  Email to check.
	 * @return  boolean
	 */
	public function is_admin_from( $email ) {

		return get_option( 'admin_email' ) == $email;

	}

	/**
	 * Register Activation
	 *
	 * Called when plugin is activated. Not called when plugin is auto-updated.
	 *
	 * @since  1.1
	 */
	public static function activate() {

		// Copy values from original WP MailFrom if present and plugin options not yet set.
		// http://wordpress.org/plugins/wp-mailfrom/
		$name = get_option( 'site_mail_from_name', '' );
		$email = get_option( 'site_mail_from_email', '' );
		$new_name = get_option( 'wp_mailfrom_ii_name', '' );
		$new_email = get_option( 'wp_mailfrom_ii_email', '' );

		if ( ! empty( $name ) && empty( $new_name ) ) {
			$name_updated = add_option( 'wp_mailfrom_ii_name', $name );
		}

		if ( ! empty( $email ) && empty( $new_email ) ) {
			$email_updated = add_option( 'wp_mailfrom_ii_email', $email );
		}

	}

}
